packages/hoppscotch-js-sandbox/src/utils.ts

```diff
@@ -1,89 +1,439 @@
-import * as O from "fp-ts/Option"
+import { parseTemplateStringE } from "@hoppscotch/data"
 import * as E from "fp-ts/Either"
-import * as QuickJS from "quickjs-emscripten"
-import { TestResult } from "./test-runner"
+import * as O from "fp-ts/Option"
+import { pipe } from "fp-ts/lib/function"
+import { cloneDeep } from "lodash-es"
 
-export function marshalObjectToVM(
-  vm: QuickJS.QuickJSVm,
-  obj: object
-): E.Either<string, QuickJS.QuickJSHandle> {
-  let jsonString
+import {
+  GlobalEnvItem,
+  SelectedEnvItem,
+  TestDescriptor,
+  TestResult,
+} from "./types"
 
-  try {
-    jsonString = JSON.stringify(obj)
-  } catch (e) {
-    return E.left("Marshaling stringification failed")
+const getEnv = (envName: string, envs: TestResult["envs"]) => {
+  return O.fromNullable(
+    envs.selected.find((x: SelectedEnvItem) => x.key === envName) ??
+      envs.global.find((x: GlobalEnvItem) => x.key === envName)
+  )
+}
+
+// Compiles shared scripting API methods for use in both pre and post request scripts
+const getSharedMethods = (envs: TestResult["envs"]) => {
```

Comment:
The `envs` argument received here is a deep copy. From this point, we rely on mutation since most of them happen in the API methods, which, while invoked, can't return the updated value to the surrounding context.
For instance, the environment gets updated in the `setEnvFn` method.

This is the case with `createExpectation` function as well (`currTestStack`).